snorkel/udf.py

```diff
@@ -92,54 +90,64 @@ def apply_mt(self, xs, parallelism, **kwargs):
         for x in xs:
             in_queue.put(x)
 
-        # If the UDF has a reduce step, we collect the output of apply in a Queue
-        out_queue = None
-        if hasattr(self.udf_class, 'reduce'):
-            out_queue = JoinableQueue()
+        # If the UDF has a reduce step, we collect the output of apply in a
+        # Queue. This is also used to track progress via the the UDF sentinel
+        out_queue = JoinableQueue()
+
+        # Keep track of progress counts
+        total_count = in_queue.qsize()
+        count = 0
 
         # Start UDF Processes
         for i in range(parallelism):
-            udf              = self.udf_class(in_queue=in_queue, out_queue=out_queue, **self.udf_init_kwargs)
+            udf = self.udf_class(in_queue=in_queue, out_queue=out_queue,
+                add_to_session=(self.reducer is None), **self.udf_init_kwargs)
             udf.apply_kwargs = kwargs
             self.udfs.append(udf)
 
         # Start the UDF processes, and then join on their completion
         for udf in self.udfs:
             udf.start()
 
-        # If there is a reduce step, do now on this thread
-        if hasattr(self.udf_class, 'reduce'):
-            while any([udf.is_alive() for udf in self.udfs]):
-                while True:
-                    try:
-                        y = out_queue.get(True, QUEUE_TIMEOUT)
-                        self.reducer.reduce(y, **kwargs)
-                        out_queue.task_done()
-                    except Empty:
-                        break
-                self.reducer.session.commit()
-            self.reducer.session.close()
+        while any([udf.is_alive() for udf in self.udfs]) and count < total_count:
+            y = out_queue.get()
```

Comment:
What's the reason behind getting rid of `QUEUE_TIMEOUT` here?  If no specific one, let's at least leave in as a kwarg?

Reply:
In the case where there is a reduce step, the timeout places an arbitrary restriction/estimate on the completion time of the user-defined function (which in our case, was too low). If the UDF isn't fast enough to fill up the out_queue, this will just busy-loop at which point, why not leave that up to the queue to decide… seems like `get()` is cleaner. Let me know what you think.

Reply:
No makes sense- so all good by me then!